File: src/main/java/uk/co/eelpieconsulting/osm/nominatim/elasticsearch/profiles/CountryCityTownSuburb.java
package uk.co.eelpieconsulting.osm.nominatim.elasticsearch.profiles;

import org.elasticsearch.index.query.BoolQueryBuilder;
import org.elasticsearch.index.query.QueryBuilder;

import static org.elasticsearch.index.query.QueryBuilders.boolQuery;
import static org.elasticsearch.index.query.QueryBuilders.termQuery;

public class CountryCityTownSuburb implements Profile {

    public static final String COUNTRY_CITY_TOWN_SUBURB = "countryCityTownSuburb";
    private static final String TAGS = "tags";

    public String getName() {
        return COUNTRY_CITY_TOWN_SUBURB;
    }

    public BoolQueryBuilder getQuery() {
        QueryBuilder isCountry = termQuery(TAGS, "place|country");
        QueryBuilder isCity = termQuery(TAGS, "place|city");
        QueryBuilder isCounty = termQuery(TAGS, "place|county");
        QueryBuilder isTown = termQuery(TAGS, "place|town");
        QueryBuilder isSuburb = termQuery(TAGS, "place|suburb");
        QueryBuilder isNationalPark = termQuery(TAGS, "boundary|national_park");
        QueryBuilder isLeisurePark = termQuery(TAGS, "leisure|park");
        QueryBuilder isLeisureCommon = termQuery(TAGS, "leisure|common");
        QueryBuilder isPeak = termQuery(TAGS, "natural|peak");
        QueryBuilder isIsland = termQuery(TAGS, "place|island");
        QueryBuilder isVillage = termQuery(TAGS, "place|village");
        QueryBuilder isBoundary = termQuery(TAGS, "boundary|administrative");
        QueryBuilder isAdminLevelFour = termQuery("adminLevel", "4");
        QueryBuilder isAdminLevelFourBoundary = boolQuery().must(isBoundary).must(isAdminLevelFour);
        QueryBuilder isAdminLevelSix = termQuery("adminLevel", "6");
        QueryBuilder isAdminLevelSixBoundary = boolQuery().must(isBoundary).must(isAdminLevelSix);

        return boolQuery().minimumNumberShouldMatch(1).
                should(isCountry).boost(10).
                should(isCity).boost(8).
                should(isNationalPark).boost(8).
                should(isAdminLevelFourBoundary).boost(6).
                should(isAdminLevelSixBoundary).boost(5).
                should(isCounty).boost(4).
                should(isTown).boost(3).
                should(isPeak).boost(3).
                should(isIsland).
                should(isLeisurePark).
                should(isLeisureCommon).
                should(isVillage).
                should(isSuburb);
    }

}

File: src/main/kotlin/uk/co/eelpieconsulting/osm/nominatim/elasticsearch/profiles/CountryCityTownSuburb.kt
package uk.co.eelpieconsulting.osm.nominatim.elasticsearch.profiles

import org.elasticsearch.index.query.BoolQueryBuilder
import org.elasticsearch.index.query.QueryBuilders.boolQuery

class CountryCityTownSuburb : Profile {

    override fun getName(): String {
        return "countryCityTownSuburb"
    }

    override fun getQuery(): BoolQueryBuilder {
        val isCountry = requiringTag("place|country")
        val isCity = requiringTag("place|city")
        val isCounty = requiringTag("place|county")
        val isTown = requiringTag("place|town")
        val isSuburb = requiringTag("place|suburb")
        val isNationalPark = requiringTag("boundary|national_park")
        val isLeisurePark = requiringTag("leisure|park")
        val isLeisureCommon = requiringTag("leisure|common")
        val isPeak = requiringTag("natural|peak")
        val isIsland = requiringTag("place|island")
        val isVillage = requiringTag("place|village")
        val isBoundary = requiringTag("boundary|administrative")
        val isAdminLevelFour = requiringAdminLevel("4")
        val isAdminLevelFourBoundary = boolQuery().must(isBoundary).must(isAdminLevelFour)
        val isAdminLevelSix = requiringAdminLevel("6")
        val isAdminLevelSixBoundary = boolQuery().must(isBoundary).must(isAdminLevelSix)

        return boolQuery().minimumShouldMatch(1).should(isCountry).boost(10f).should(isCity).boost(8f)
            .should(isNationalPark).boost(8f).should(isAdminLevelFourBoundary).boost(6f).should(isAdminLevelSixBoundary)
            .boost(5f).should(isCounty).boost(4f).should(isTown).boost(3f).should(isPeak).boost(3f).should(isIsland)
            .should(isLeisurePark).should(isLeisureCommon).should(isVillage).should(isSuburb)
    }

}
